// cosmossdkjs/module/proto/osmosis/txfees/v1beta1/genesis.js
import { FeeToken } from "./feetoken";
import * as _m0 from "protobufjs/minimal";
import { isSet } from "@osmonauts/helpers";
/** GenesisState defines the txfees module's genesis state. */

function createBaseGenesisState() {
  return {
    basedenom: "",
    feetokens: []
  };
}

export const GenesisState = {
  encode(message, writer = _m0.Writer.create()) {
    if (message.basedenom !== "") {
      writer.uint32(10).string(message.basedenom);
    }

    for (const v of message.feetokens) {
      FeeToken.encode(v, writer.uint32(18).fork()).ldelim();
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseGenesisState();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.basedenom = reader.string();
          break;

        case 2:
          message.feetokens.push(FeeToken.decode(reader, reader.uint32()));
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object) {
    return {
      basedenom: isSet(object.basedenom) ? String(object.basedenom) : "",
      feetokens: Array.isArray(object?.feetokens) ? object.feetokens.map(e => FeeToken.fromJSON(e)) : []
    };
  },

  toJSON(message) {
    const obj = {};
    message.basedenom !== undefined && (obj.basedenom = message.basedenom);

    if (message.feetokens) {
      obj.feetokens = message.feetokens.map(e => e ? FeeToken.toJSON(e) : undefined);
    } else {
      obj.feetokens = [];
    }

    return obj;
  },

  fromPartial(object) {
    const message = createBaseGenesisState();
    message.basedenom = object.basedenom ?? "";
    message.feetokens = object.feetokens?.map(e => FeeToken.fromPartial(e)) || [];
    return message;
  }

};
