// cosmossdkjs/module/proto/osmosis/txfees/v1beta1/query.lcd.js
import { LCDClient } from "@osmonauts/lcd";
export class LCDQueryClient extends LCDClient {
  constructor({
    restEndpoint
  }) {
    super({
      restEndpoint
    });
  }
  /* FeeTokens returns a list of all the whitelisted fee tokens and their
  corresponding pools It does not include the BaseDenom, which has its own
  query endpoint */


  async feeTokens(_params = {}) {
    const endpoint = `osmosis/txfees/v1beta1/fee_tokens`;
    return await this.request(endpoint);
  }
  /* DenomSpotPrice */


  async denomSpotPrice(params) {
    const options = {
      params: {}
    };

    if (typeof params?.denom !== "undefined") {
      options.params.denom = params.denom;
    }

    const endpoint = `osmosis/txfees/v1beta1/spot_price_by_denom`;
    return await this.request(endpoint, options);
  }
  /* DenomPoolId */


  async denomPoolId(params) {
    const endpoint = `osmosis/txfees/v1beta1/denom_pool_id/${params.denom}`;
    return await this.request(endpoint);
  }
  /* BaseDenom */


  async baseDenom(_params = {}) {
    const endpoint = `osmosis/txfees/v1beta1/base_denom`;
    return await this.request(endpoint);
  }

}
