// chain-registry/main/mainnet/stratos/assets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var info = {
  $schema: '../assetlist.schema.json',
  chain_name: 'stratos',
  assets: [{
    description: 'STOS coin is the token for the Stratos (STOS) cosmos based blockchain',
    denom_units: [{
      denom: 'wei',
      exponent: 0
    }, {
      denom: 'gwei',
      exponent: 9
    }, {
      denom: 'stos',
      exponent: 18
    }],
    base: 'wei',
    name: 'Stratos',
    display: 'stos',
    symbol: 'STOS',
    logo_URIs: {
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/stratos/images/stratos.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/stratos/images/stratos.svg'
    },
    coingecko_id: 'stratos',
    images: [{
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/stratos/images/stratos.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/stratos/images/stratos.svg'
    }]
  }]
};
var _default = exports["default"] = info;
