// cosmossdkjs/module/proto/cosmos/base/reflection/v1beta1/reflection.js
import * as _m0 from "protobufjs/minimal";
import { isSet } from "@osmonauts/helpers";
/** ListAllInterfacesRequest is the request type of the ListAllInterfaces RPC. */

function createBaseListAllInterfacesRequest() {
  return {};
}

export const ListAllInterfacesRequest = {
  encode(_, writer = _m0.Writer.create()) {
    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseListAllInterfacesRequest();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(_) {
    return {};
  },

  toJSON(_) {
    const obj = {};
    return obj;
  },

  fromPartial(_) {
    const message = createBaseListAllInterfacesRequest();
    return message;
  }

};

function createBaseListAllInterfacesResponse() {
  return {
    interfaceNames: []
  };
}

export const ListAllInterfacesResponse = {
  encode(message, writer = _m0.Writer.create()) {
    for (const v of message.interfaceNames) {
      writer.uint32(10).string(v);
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseListAllInterfacesResponse();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.interfaceNames.push(reader.string());
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object) {
    return {
      interfaceNames: Array.isArray(object?.interfaceNames) ? object.interfaceNames.map(e => String(e)) : []
    };
  },

  toJSON(message) {
    const obj = {};

    if (message.interfaceNames) {
      obj.interfaceNames = message.interfaceNames.map(e => e);
    } else {
      obj.interfaceNames = [];
    }

    return obj;
  },

  fromPartial(object) {
    const message = createBaseListAllInterfacesResponse();
    message.interfaceNames = object.interfaceNames?.map(e => e) || [];
    return message;
  }

};

function createBaseListImplementationsRequest() {
  return {
    interfaceName: ""
  };
}

export const ListImplementationsRequest = {
  encode(message, writer = _m0.Writer.create()) {
    if (message.interfaceName !== "") {
      writer.uint32(10).string(message.interfaceName);
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseListImplementationsRequest();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.interfaceName = reader.string();
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object) {
    return {
      interfaceName: isSet(object.interfaceName) ? String(object.interfaceName) : ""
    };
  },

  toJSON(message) {
    const obj = {};
    message.interfaceName !== undefined && (obj.interfaceName = message.interfaceName);
    return obj;
  },

  fromPartial(object) {
    const message = createBaseListImplementationsRequest();
    message.interfaceName = object.interfaceName ?? "";
    return message;
  }

};

function createBaseListImplementationsResponse() {
  return {
    implementationMessageNames: []
  };
}

export const ListImplementationsResponse = {
  encode(message, writer = _m0.Writer.create()) {
    for (const v of message.implementationMessageNames) {
      writer.uint32(10).string(v);
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseListImplementationsResponse();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.implementationMessageNames.push(reader.string());
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object) {
    return {
      implementationMessageNames: Array.isArray(object?.implementationMessageNames) ? object.implementationMessageNames.map(e => String(e)) : []
    };
  },

  toJSON(message) {
    const obj = {};

    if (message.implementationMessageNames) {
      obj.implementationMessageNames = message.implementationMessageNames.map(e => e);
    } else {
      obj.implementationMessageNames = [];
    }

    return obj;
  },

  fromPartial(object) {
    const message = createBaseListImplementationsResponse();
    message.implementationMessageNames = object.implementationMessageNames?.map(e => e) || [];
    return message;
  }

};
