// cosmossdkjs/module/clients/client.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { DirectSecp256k1HdWallet } from "@cosmjs/proto-signing";
import { createProtobufRpcClient, QueryClient, SigningStargateClient } from "@cosmjs/stargate";
import { SigningCosmWasmClient } from "@cosmjs/cosmwasm-stargate";
import { Tendermint34Client } from "@cosmjs/tendermint-rpc";
import { PersistenceWalletOptions, PersistenceLocalConfig } from "../config/config";
import { createRPCQueryClient } from "../proto/cosmwasm/rpc.query";
export class CosmosChainClient {
  constructor(mnemonic, wallet, signer, wasm, core, queryclient, config) {
    _defineProperty(this, "wallet", void 0);

    _defineProperty(this, "signer", void 0);

    _defineProperty(this, "mnemonic", void 0);

    _defineProperty(this, "config", void 0);

    _defineProperty(this, "core", void 0);

    _defineProperty(this, "wasm", void 0);

    _defineProperty(this, "query", void 0);

    this.mnemonic = mnemonic;
    this.config = config;
    this.wallet = wallet;
    this.signer = signer;
    this.core = core;
    this.wasm = wasm;
    this.query = queryclient;
  }

  static async init(mnemonic, chainConfig, wallet_options, signer) {
    //wallet
    const config = chainConfig || PersistenceLocalConfig;
    const act_wallet_options = wallet_options || PersistenceWalletOptions;
    const wallet = await DirectSecp256k1HdWallet.fromMnemonic(mnemonic, act_wallet_options);
    const clientsigner = signer || wallet; //signing clients

    const core = await SigningStargateClient.connectWithSigner(config.rpc, clientsigner);
    const wasm = await SigningCosmWasmClient.connectWithSigner(config.rpc, clientsigner); //query client

    const tendermintClient = await Tendermint34Client.connect(config.rpc);
    const queryClient = new QueryClient(tendermintClient);
    const rpc = await createProtobufRpcClient(queryClient);
    const query = await createRPCQueryClient({
      rpc: rpc
    });
    return new CosmosChainClient(mnemonic, wallet, clientsigner, wasm, core, query, config);
  }

}
