// @cosmos-client/core/esm/rest/bank/module.js
import { QueryApi } from '../../openapi/api';
export function allBalances(sdk, address, paginationKey, paginationOffset, paginationLimit, paginationCountTotal) {
    return new QueryApi(undefined, sdk.url).allBalances(address, paginationKey, paginationOffset?.toString(), paginationLimit?.toString(), paginationCountTotal);
}
export function balance(sdk, address, denom) {
    return new QueryApi(undefined, sdk.url).balance(address, denom);
}
export function params(sdk) {
    return new QueryApi(undefined, sdk.url).bankParams();
}
export function supplyOf(sdk, denom) {
    return new QueryApi(undefined, sdk.url).supplyOf(denom);
}
export function totalSupply(sdk) {
    return new QueryApi(undefined, sdk.url).totalSupply();
}
export function denomOwners(sdk, denom) {
    return new QueryApi(undefined, sdk.url).denomOwners(denom);
}
export function denomMetadata(sdk, denom) {
    return new QueryApi(undefined, sdk.url).denomMetadata(denom);
}
export function denomsMetadata(sdk) {
    return new QueryApi(undefined, sdk.url).denomsMetadata();
}
