// cosmossdkjs/main/clients/client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CosmosChainClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _protoSigning = require("@cosmjs/proto-signing");

var _stargate = require("@cosmjs/stargate");

var _cosmwasmStargate = require("@cosmjs/cosmwasm-stargate");

var _tendermintRpc = require("@cosmjs/tendermint-rpc");

var _config = require("../config/config");

var _rpc = require("../proto/cosmwasm/rpc.query");

var CosmosChainClient = /*#__PURE__*/function () {
  function CosmosChainClient(mnemonic, wallet, signer, wasm, core, queryclient, config) {
    (0, _classCallCheck2["default"])(this, CosmosChainClient);
    (0, _defineProperty2["default"])(this, "wallet", void 0);
    (0, _defineProperty2["default"])(this, "signer", void 0);
    (0, _defineProperty2["default"])(this, "mnemonic", void 0);
    (0, _defineProperty2["default"])(this, "config", void 0);
    (0, _defineProperty2["default"])(this, "core", void 0);
    (0, _defineProperty2["default"])(this, "wasm", void 0);
    (0, _defineProperty2["default"])(this, "query", void 0);
    this.mnemonic = mnemonic;
    this.config = config;
    this.wallet = wallet;
    this.signer = signer;
    this.core = core;
    this.wasm = wasm;
    this.query = queryclient;
  }

  (0, _createClass2["default"])(CosmosChainClient, null, [{
    key: "init",
    value: function () {
      var _init = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(mnemonic, chainConfig, wallet_options, signer) {
        var config, act_wallet_options, wallet, clientsigner, core, wasm, tendermintClient, queryClient, rpc, query;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                //wallet
                config = chainConfig || _config.PersistenceLocalConfig;
                act_wallet_options = wallet_options || _config.PersistenceWalletOptions;
                _context.next = 4;
                return _protoSigning.DirectSecp256k1HdWallet.fromMnemonic(mnemonic, act_wallet_options);

              case 4:
                wallet = _context.sent;
                clientsigner = signer || wallet; //signing clients

                _context.next = 8;
                return _stargate.SigningStargateClient.connectWithSigner(config.rpc, clientsigner);

              case 8:
                core = _context.sent;
                _context.next = 11;
                return _cosmwasmStargate.SigningCosmWasmClient.connectWithSigner(config.rpc, clientsigner);

              case 11:
                wasm = _context.sent;
                _context.next = 14;
                return _tendermintRpc.Tendermint34Client.connect(config.rpc);

              case 14:
                tendermintClient = _context.sent;
                queryClient = new _stargate.QueryClient(tendermintClient);
                _context.next = 18;
                return (0, _stargate.createProtobufRpcClient)(queryClient);

              case 18:
                rpc = _context.sent;
                _context.next = 21;
                return (0, _rpc.createRPCQueryClient)({
                  rpc: rpc
                });

              case 21:
                query = _context.sent;
                return _context.abrupt("return", new CosmosChainClient(mnemonic, wallet, clientsigner, wasm, core, query, config));

              case 23:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function init(_x, _x2, _x3, _x4) {
        return _init.apply(this, arguments);
      }

      return init;
    }()
  }]);
  return CosmosChainClient;
}();

exports.CosmosChainClient = CosmosChainClient;
