// cosmossdkjs/main/utils/cosmos/authz.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.execute_cosmos_authz_exec = execute_cosmos_authz_exec;
exports.execute_cosmos_authz_grant = execute_cosmos_authz_grant;
exports.execute_cosmos_authz_revoke = execute_cosmos_authz_revoke;
exports.query_cosmos_authz_grants = query_cosmos_authz_grants;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

// Supported query interfaces - types/proto/cosmos/authz/v1beta1/query.rpc.query.d.ts
// Supported tranasction interfaces - types/proto/cosmos/authz/v1beta1/tx.rpc.msg.d.ts
//---------------------------------------------
// COSMOS::AUTHZ MODULE - QUERY HELPERS
// --------------------------------------------
function query_cosmos_authz_grants(_x, _x2, _x3, _x4, _x5, _x6) {
  return _query_cosmos_authz_grants.apply(this, arguments);
} //---------------------------------------------
// COSMOS::AUTHZ MODULE - TRANSACTION HELPERS
// --------------------------------------------
// MsgGrant


function _query_cosmos_authz_grants() {
  _query_cosmos_authz_grants = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client, granter, grantee, msgTypeUrl, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return client.query.cosmos.authz.v1beta1.grants(_proto.cosmos.authz.v1beta1.QueryGrantsRequest.fromPartial({
              granter: granter,
              grantee: grantee,
              msgTypeUrl: msgTypeUrl,
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _query_cosmos_authz_grants.apply(this, arguments);
}

function execute_cosmos_authz_grant(_x7, _x8, _x9, _x10) {
  return _execute_cosmos_authz_grant.apply(this, arguments);
} // MsgExec


function _execute_cosmos_authz_grant() {
  _execute_cosmos_authz_grant = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(granter, grantee, authorization, expiration) {
    var msg;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            msg = _proto.cosmos.authz.v1beta1.MsgGrant.fromPartial({
              granter: granter,
              grantee: grantee,
              grant: _proto.cosmos.authz.v1beta1.Grant.fromPartial({
                authorization: authorization,
                expiration: expiration
              })
            });
            return _context2.abrupt("return", msg);

          case 2:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _execute_cosmos_authz_grant.apply(this, arguments);
}

function execute_cosmos_authz_exec(_x11, _x12) {
  return _execute_cosmos_authz_exec.apply(this, arguments);
} // MsgRevoke


function _execute_cosmos_authz_exec() {
  _execute_cosmos_authz_exec = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(grantee, msgs) {
    var msg;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            msg = _proto.cosmos.authz.v1beta1.MsgExec.fromPartial({
              grantee: grantee,
              msgs: msgs
            });
            return _context3.abrupt("return", msg);

          case 2:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _execute_cosmos_authz_exec.apply(this, arguments);
}

function execute_cosmos_authz_revoke(_x13, _x14, _x15) {
  return _execute_cosmos_authz_revoke.apply(this, arguments);
}

function _execute_cosmos_authz_revoke() {
  _execute_cosmos_authz_revoke = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(granter, grantee, msgTypeUrl) {
    var msg;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            msg = _proto.cosmos.authz.v1beta1.MsgRevoke.fromPartial({
              granter: granter,
              grantee: grantee,
              msgTypeUrl: msgTypeUrl
            });
            return _context4.abrupt("return", msg);

          case 2:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _execute_cosmos_authz_revoke.apply(this, arguments);
}
