// cosmossdkjs/main/utils/cosmos/auth.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.query_cosmos_auth_account = query_cosmos_auth_account;
exports.query_cosmos_auth_accounts = query_cosmos_auth_accounts;
exports.query_cosmos_auth_params = query_cosmos_auth_params;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

//Supported query interfaces - types/proto/cosmos/auth/v1beta1/query.rpc.query.d.ts
//---------------------------------------------
// COSMOS::AUTH MODULE - QUERY HELPERS
// --------------------------------------------
function query_cosmos_auth_account(_x, _x2) {
  return _query_cosmos_auth_account.apply(this, arguments);
}

function _query_cosmos_auth_account() {
  _query_cosmos_auth_account = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client, address) {
    var response;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return client.query.cosmos.auth.v1beta1.account(_proto.cosmos.auth.v1beta1.QueryAccountRequest.fromPartial({
              address: address
            }));

          case 2:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _query_cosmos_auth_account.apply(this, arguments);
}

function query_cosmos_auth_accounts(_x3, _x4, _x5) {
  return _query_cosmos_auth_accounts.apply(this, arguments);
}

function _query_cosmos_auth_accounts() {
  _query_cosmos_auth_accounts = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return client.query.cosmos.auth.v1beta1.accounts(_proto.cosmos.auth.v1beta1.QueryAccountsRequest.fromPartial({
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _query_cosmos_auth_accounts.apply(this, arguments);
}

function query_cosmos_auth_params(_x6) {
  return _query_cosmos_auth_params.apply(this, arguments);
}

function _query_cosmos_auth_params() {
  _query_cosmos_auth_params = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client) {
    var response;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return client.query.cosmos.auth.v1beta1.params(_proto.cosmos.auth.v1beta1.QueryParamsRequest.fromPartial({}));

          case 2:
            response = _context3.sent;
            return _context3.abrupt("return", response);

          case 4:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _query_cosmos_auth_params.apply(this, arguments);
}
