// cosmossdkjs/main/utils/cosmos/distribution.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.execute_cosmos_distribution_fund_community_pool = execute_cosmos_distribution_fund_community_pool;
exports.execute_cosmos_distribution_set_withdraw_address = execute_cosmos_distribution_set_withdraw_address;
exports.execute_cosmos_distribution_withdraw_delegator_reward = execute_cosmos_distribution_withdraw_delegator_reward;
exports.execute_cosmos_distribution_withdraw_validator_commission = execute_cosmos_distribution_withdraw_validator_commission;
exports.query_cosmos_distribution_community_pool = query_cosmos_distribution_community_pool;
exports.query_cosmos_distribution_delegationRewards = query_cosmos_distribution_delegationRewards;
exports.query_cosmos_distribution_delegation_total_rewards = query_cosmos_distribution_delegation_total_rewards;
exports.query_cosmos_distribution_delegator_validators = query_cosmos_distribution_delegator_validators;
exports.query_cosmos_distribution_delegator_withdraw_address = query_cosmos_distribution_delegator_withdraw_address;
exports.query_cosmos_distribution_params = query_cosmos_distribution_params;
exports.query_cosmos_distribution_validator_commission = query_cosmos_distribution_validator_commission;
exports.query_cosmos_distribution_validator_outstanding_rewards = query_cosmos_distribution_validator_outstanding_rewards;
exports.query_cosmos_distribution_validator_slashes = query_cosmos_distribution_validator_slashes;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

// Supported query interfaces - types/proto/cosmos/distribution/v1beta1/query.rpc.query.d.ts
// Supported tranasction interfaces - types/proto/cosmos/distribution/v1beta1/tx.rpc.msg.d.ts
//---------------------------------------------
// COSMOS::DISTRIBUTION MODULE - QUERY HELPERS
// --------------------------------------------
function query_cosmos_distribution_params(_x) {
  return _query_cosmos_distribution_params.apply(this, arguments);
}

function _query_cosmos_distribution_params() {
  _query_cosmos_distribution_params = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client) {
    var response;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return client.query.cosmos.distribution.v1beta1.params(_proto.cosmos.distribution.v1beta1.QueryParamsRequest.fromPartial({}));

          case 2:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _query_cosmos_distribution_params.apply(this, arguments);
}

function query_cosmos_distribution_validator_outstanding_rewards(_x2, _x3) {
  return _query_cosmos_distribution_validator_outstanding_rewards.apply(this, arguments);
}

function _query_cosmos_distribution_validator_outstanding_rewards() {
  _query_cosmos_distribution_validator_outstanding_rewards = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client, validatorAddress) {
    var response;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return client.query.cosmos.distribution.v1beta1.validatorOutstandingRewards(_proto.cosmos.distribution.v1beta1.QueryValidatorOutstandingRewardsRequest.fromPartial({
              validatorAddress: validatorAddress
            }));

          case 2:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _query_cosmos_distribution_validator_outstanding_rewards.apply(this, arguments);
}

function query_cosmos_distribution_validator_commission(_x4, _x5) {
  return _query_cosmos_distribution_validator_commission.apply(this, arguments);
}

function _query_cosmos_distribution_validator_commission() {
  _query_cosmos_distribution_validator_commission = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client, validatorAddress) {
    var response;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return client.query.cosmos.distribution.v1beta1.validatorCommission(_proto.cosmos.distribution.v1beta1.QueryValidatorCommissionRequest.fromPartial({
              validatorAddress: validatorAddress
            }));

          case 2:
            response = _context3.sent;
            return _context3.abrupt("return", response);

          case 4:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _query_cosmos_distribution_validator_commission.apply(this, arguments);
}

function query_cosmos_distribution_validator_slashes(_x6, _x7, _x8, _x9, _x10, _x11) {
  return _query_cosmos_distribution_validator_slashes.apply(this, arguments);
}

function _query_cosmos_distribution_validator_slashes() {
  _query_cosmos_distribution_validator_slashes = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(client, validatorAddress, startingHeight, endingHeight, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return client.query.cosmos.distribution.v1beta1.validatorSlashes(_proto.cosmos.distribution.v1beta1.QueryValidatorSlashesRequest.fromPartial({
              validatorAddress: validatorAddress,
              startingHeight: startingHeight,
              endingHeight: endingHeight,
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context4.sent;
            return _context4.abrupt("return", response);

          case 4:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _query_cosmos_distribution_validator_slashes.apply(this, arguments);
}

function query_cosmos_distribution_delegationRewards(_x12, _x13, _x14) {
  return _query_cosmos_distribution_delegationRewards.apply(this, arguments);
}

function _query_cosmos_distribution_delegationRewards() {
  _query_cosmos_distribution_delegationRewards = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(client, delegatorAddress, validatorAddress) {
    var response;
    return _regenerator["default"].wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.next = 2;
            return client.query.cosmos.distribution.v1beta1.delegationRewards(_proto.cosmos.distribution.v1beta1.QueryDelegationRewardsRequest.fromPartial({
              delegatorAddress: delegatorAddress,
              validatorAddress: validatorAddress
            }));

          case 2:
            response = _context5.sent;
            return _context5.abrupt("return", response);

          case 4:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _query_cosmos_distribution_delegationRewards.apply(this, arguments);
}

function query_cosmos_distribution_delegation_total_rewards(_x15, _x16) {
  return _query_cosmos_distribution_delegation_total_rewards.apply(this, arguments);
}

function _query_cosmos_distribution_delegation_total_rewards() {
  _query_cosmos_distribution_delegation_total_rewards = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(client, delegatorAddress) {
    var response;
    return _regenerator["default"].wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            _context6.next = 2;
            return client.query.cosmos.distribution.v1beta1.delegationTotalRewards(_proto.cosmos.distribution.v1beta1.QueryDelegationTotalRewardsRequest.fromPartial({
              delegatorAddress: delegatorAddress
            }));

          case 2:
            response = _context6.sent;
            return _context6.abrupt("return", response);

          case 4:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));
  return _query_cosmos_distribution_delegation_total_rewards.apply(this, arguments);
}

function query_cosmos_distribution_delegator_validators(_x17, _x18) {
  return _query_cosmos_distribution_delegator_validators.apply(this, arguments);
}

function _query_cosmos_distribution_delegator_validators() {
  _query_cosmos_distribution_delegator_validators = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(client, delegatorAddress) {
    var response;
    return _regenerator["default"].wrap(function _callee7$(_context7) {
      while (1) {
        switch (_context7.prev = _context7.next) {
          case 0:
            _context7.next = 2;
            return client.query.cosmos.distribution.v1beta1.delegatorValidators(_proto.cosmos.distribution.v1beta1.QueryDelegatorValidatorsRequest.fromPartial({
              delegatorAddress: delegatorAddress
            }));

          case 2:
            response = _context7.sent;
            return _context7.abrupt("return", response);

          case 4:
          case "end":
            return _context7.stop();
        }
      }
    }, _callee7);
  }));
  return _query_cosmos_distribution_delegator_validators.apply(this, arguments);
}

function query_cosmos_distribution_delegator_withdraw_address(_x19, _x20) {
  return _query_cosmos_distribution_delegator_withdraw_address.apply(this, arguments);
}

function _query_cosmos_distribution_delegator_withdraw_address() {
  _query_cosmos_distribution_delegator_withdraw_address = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(client, delegatorAddress) {
    var response;
    return _regenerator["default"].wrap(function _callee8$(_context8) {
      while (1) {
        switch (_context8.prev = _context8.next) {
          case 0:
            _context8.next = 2;
            return client.query.cosmos.distribution.v1beta1.delegatorWithdrawAddress(_proto.cosmos.distribution.v1beta1.QueryDelegatorWithdrawAddressRequest.fromPartial({
              delegatorAddress: delegatorAddress
            }));

          case 2:
            response = _context8.sent;
            return _context8.abrupt("return", response);

          case 4:
          case "end":
            return _context8.stop();
        }
      }
    }, _callee8);
  }));
  return _query_cosmos_distribution_delegator_withdraw_address.apply(this, arguments);
}

function query_cosmos_distribution_community_pool(_x21) {
  return _query_cosmos_distribution_community_pool.apply(this, arguments);
} //---------------------------------------------
// COSMOS::DISTRIBUTION MODULE - TRANSACTION HELPERS
// --------------------------------------------
//MsgSetWithdrawAddress


function _query_cosmos_distribution_community_pool() {
  _query_cosmos_distribution_community_pool = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(client) {
    var response;
    return _regenerator["default"].wrap(function _callee9$(_context9) {
      while (1) {
        switch (_context9.prev = _context9.next) {
          case 0:
            _context9.next = 2;
            return client.query.cosmos.distribution.v1beta1.communityPool(_proto.cosmos.distribution.v1beta1.QueryCommunityPoolRequest.fromPartial({}));

          case 2:
            response = _context9.sent;
            return _context9.abrupt("return", response);

          case 4:
          case "end":
            return _context9.stop();
        }
      }
    }, _callee9);
  }));
  return _query_cosmos_distribution_community_pool.apply(this, arguments);
}

function execute_cosmos_distribution_set_withdraw_address(_x22, _x23) {
  return _execute_cosmos_distribution_set_withdraw_address.apply(this, arguments);
} // MsgWithdrawDelegatorReward


function _execute_cosmos_distribution_set_withdraw_address() {
  _execute_cosmos_distribution_set_withdraw_address = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(delegatorAddress, withdrawAddress) {
    var msg;
    return _regenerator["default"].wrap(function _callee10$(_context10) {
      while (1) {
        switch (_context10.prev = _context10.next) {
          case 0:
            msg = _proto.cosmos.distribution.v1beta1.MsgSetWithdrawAddress.fromPartial({
              delegatorAddress: delegatorAddress,
              withdrawAddress: withdrawAddress
            });
            return _context10.abrupt("return", msg);

          case 2:
          case "end":
            return _context10.stop();
        }
      }
    }, _callee10);
  }));
  return _execute_cosmos_distribution_set_withdraw_address.apply(this, arguments);
}

function execute_cosmos_distribution_withdraw_delegator_reward(_x24, _x25) {
  return _execute_cosmos_distribution_withdraw_delegator_reward.apply(this, arguments);
} //MsgWithdrawValidatorCommission


function _execute_cosmos_distribution_withdraw_delegator_reward() {
  _execute_cosmos_distribution_withdraw_delegator_reward = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(delegatorAddress, validatorAddress) {
    var msg;
    return _regenerator["default"].wrap(function _callee11$(_context11) {
      while (1) {
        switch (_context11.prev = _context11.next) {
          case 0:
            msg = _proto.cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward.fromPartial({
              delegatorAddress: delegatorAddress,
              validatorAddress: validatorAddress
            });
            return _context11.abrupt("return", msg);

          case 2:
          case "end":
            return _context11.stop();
        }
      }
    }, _callee11);
  }));
  return _execute_cosmos_distribution_withdraw_delegator_reward.apply(this, arguments);
}

function execute_cosmos_distribution_withdraw_validator_commission(_x26) {
  return _execute_cosmos_distribution_withdraw_validator_commission.apply(this, arguments);
} //MsgFundCommunityPool


function _execute_cosmos_distribution_withdraw_validator_commission() {
  _execute_cosmos_distribution_withdraw_validator_commission = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(validatorAddress) {
    var msg;
    return _regenerator["default"].wrap(function _callee12$(_context12) {
      while (1) {
        switch (_context12.prev = _context12.next) {
          case 0:
            msg = _proto.cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission.fromPartial({
              validatorAddress: validatorAddress
            });
            return _context12.abrupt("return", msg);

          case 2:
          case "end":
            return _context12.stop();
        }
      }
    }, _callee12);
  }));
  return _execute_cosmos_distribution_withdraw_validator_commission.apply(this, arguments);
}

function execute_cosmos_distribution_fund_community_pool(_x27, _x28) {
  return _execute_cosmos_distribution_fund_community_pool.apply(this, arguments);
}

function _execute_cosmos_distribution_fund_community_pool() {
  _execute_cosmos_distribution_fund_community_pool = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee13(amount, depositor) {
    var msg;
    return _regenerator["default"].wrap(function _callee13$(_context13) {
      while (1) {
        switch (_context13.prev = _context13.next) {
          case 0:
            msg = _proto.cosmos.distribution.v1beta1.MsgFundCommunityPool.fromPartial({
              amount: amount,
              depositor: depositor
            });
            return _context13.abrupt("return", msg);

          case 2:
          case "end":
            return _context13.stop();
        }
      }
    }, _callee13);
  }));
  return _execute_cosmos_distribution_fund_community_pool.apply(this, arguments);
}
