// cosmossdkjs/main/utils/cosmos/feegrant.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.execute_cosmos_feegrant_grant_allowance = execute_cosmos_feegrant_grant_allowance;
exports.execute_cosmos_feegrant_revoke_allowance = execute_cosmos_feegrant_revoke_allowance;
exports.query_cosmos_feegrant_allowance = query_cosmos_feegrant_allowance;
exports.query_cosmos_feegrant_allowances = query_cosmos_feegrant_allowances;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

// Supported query interfaces - types/proto/cosmos/feegrant/v1beta1/query.rpc.query.d.ts
// Supported tranasction interfaces - types/proto/cosmos/feegrant/v1beta1/tx.rpc.msg.d.ts
//---------------------------------------------
// COSMOS::FEEGRANT MODULE - QUERY HELPERS
// --------------------------------------------
function query_cosmos_feegrant_allowance(_x, _x2, _x3) {
  return _query_cosmos_feegrant_allowance.apply(this, arguments);
}

function _query_cosmos_feegrant_allowance() {
  _query_cosmos_feegrant_allowance = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client, granter, grantee) {
    var response;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return client.query.cosmos.feegrant.v1beta1.allowance(_proto.cosmos.feegrant.v1beta1.QueryAllowanceRequest.fromPartial({
              granter: granter,
              grantee: grantee
            }));

          case 2:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _query_cosmos_feegrant_allowance.apply(this, arguments);
}

function query_cosmos_feegrant_allowances(_x4, _x5, _x6, _x7) {
  return _query_cosmos_feegrant_allowances.apply(this, arguments);
} //---------------------------------------------
// COSMOS::FEEGRANT MODULE - TRANSACTION HELPERS
// --------------------------------------------
//MsgGrantAllowance


function _query_cosmos_feegrant_allowances() {
  _query_cosmos_feegrant_allowances = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client, grantee, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return client.query.cosmos.feegrant.v1beta1.allowances(_proto.cosmos.feegrant.v1beta1.QueryAllowancesRequest.fromPartial({
              grantee: grantee,
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _query_cosmos_feegrant_allowances.apply(this, arguments);
}

function execute_cosmos_feegrant_grant_allowance(_x8, _x9, _x10) {
  return _execute_cosmos_feegrant_grant_allowance.apply(this, arguments);
} //MsgRevokeAllowance


function _execute_cosmos_feegrant_grant_allowance() {
  _execute_cosmos_feegrant_grant_allowance = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(granter, grantee, allowance) {
    var msg;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            msg = _proto.cosmos.feegrant.v1beta1.MsgGrantAllowance.fromPartial({
              granter: granter,
              grantee: grantee,
              allowance: allowance
            });
            return _context3.abrupt("return", msg);

          case 2:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _execute_cosmos_feegrant_grant_allowance.apply(this, arguments);
}

function execute_cosmos_feegrant_revoke_allowance(_x11, _x12) {
  return _execute_cosmos_feegrant_revoke_allowance.apply(this, arguments);
}

function _execute_cosmos_feegrant_revoke_allowance() {
  _execute_cosmos_feegrant_revoke_allowance = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(granter, grantee) {
    var msg;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            msg = _proto.cosmos.feegrant.v1beta1.MsgRevokeAllowance.fromPartial({
              granter: granter,
              grantee: grantee
            });
            return _context4.abrupt("return", msg);

          case 2:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _execute_cosmos_feegrant_revoke_allowance.apply(this, arguments);
}
