// cosmossdkjs/main/utils/cosmos/gov.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.execute_cosmos_gov_deposit_msg = execute_cosmos_gov_deposit_msg;
exports.execute_cosmos_gov_submit_proposal_msg = execute_cosmos_gov_submit_proposal_msg;
exports.execute_cosmos_gov_vote_msg = execute_cosmos_gov_vote_msg;
exports.execute_cosmos_gov_vote_weighted_msg = execute_cosmos_gov_vote_weighted_msg;
exports.query_cosmos_gov_deposit = query_cosmos_gov_deposit;
exports.query_cosmos_gov_deposits = query_cosmos_gov_deposits;
exports.query_cosmos_gov_params = query_cosmos_gov_params;
exports.query_cosmos_gov_proposal = query_cosmos_gov_proposal;
exports.query_cosmos_gov_proposals = query_cosmos_gov_proposals;
exports.query_cosmos_gov_tally_result = query_cosmos_gov_tally_result;
exports.query_cosmos_gov_vote = query_cosmos_gov_vote;
exports.query_cosmos_gov_votes = query_cosmos_gov_votes;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

// Supported query interfaces - types/proto/cosmos/gov/v1beta1/query.rpc.query.d.ts
// Supported tranasction interfaces - types/proto/cosmos/gov/v1beta1/tx.rpc.msg.d.ts
//---------------------------------------------
// COSMOS::GOV MODULE - QUERY HELPERS
// --------------------------------------------
function query_cosmos_gov_proposal(_x, _x2) {
  return _query_cosmos_gov_proposal.apply(this, arguments);
}

function _query_cosmos_gov_proposal() {
  _query_cosmos_gov_proposal = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client, proposalId) {
    var response;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return client.query.cosmos.gov.v1beta1.proposal(_proto.cosmos.gov.v1beta1.QueryProposalRequest.fromPartial({
              proposalId: proposalId
            }));

          case 2:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _query_cosmos_gov_proposal.apply(this, arguments);
}

function query_cosmos_gov_proposals(_x3, _x4, _x5, _x6, _x7, _x8) {
  return _query_cosmos_gov_proposals.apply(this, arguments);
}

function _query_cosmos_gov_proposals() {
  _query_cosmos_gov_proposals = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client, proposalStatus, voter, depositor, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return client.query.cosmos.gov.v1beta1.proposals(_proto.cosmos.gov.v1beta1.QueryProposalsRequest.fromPartial({
              proposalStatus: proposalStatus,
              voter: voter,
              depositor: depositor,
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _query_cosmos_gov_proposals.apply(this, arguments);
}

function query_cosmos_gov_vote(_x9, _x10, _x11) {
  return _query_cosmos_gov_vote.apply(this, arguments);
}

function _query_cosmos_gov_vote() {
  _query_cosmos_gov_vote = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client, proposalId, voter) {
    var response;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return client.query.cosmos.gov.v1beta1.vote(_proto.cosmos.gov.v1beta1.QueryVoteRequest.fromPartial({
              proposalId: proposalId,
              voter: voter
            }));

          case 2:
            response = _context3.sent;
            return _context3.abrupt("return", response);

          case 4:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _query_cosmos_gov_vote.apply(this, arguments);
}

function query_cosmos_gov_votes(_x12, _x13, _x14, _x15) {
  return _query_cosmos_gov_votes.apply(this, arguments);
}

function _query_cosmos_gov_votes() {
  _query_cosmos_gov_votes = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(client, proposalId, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return client.query.cosmos.gov.v1beta1.votes(_proto.cosmos.gov.v1beta1.QueryVotesRequest.fromPartial({
              proposalId: proposalId,
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context4.sent;
            return _context4.abrupt("return", response);

          case 4:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _query_cosmos_gov_votes.apply(this, arguments);
}

function query_cosmos_gov_params(_x16) {
  return _query_cosmos_gov_params.apply(this, arguments);
}

function _query_cosmos_gov_params() {
  _query_cosmos_gov_params = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(client) {
    var response;
    return _regenerator["default"].wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.next = 2;
            return client.query.cosmos.gov.v1beta1.params(_proto.cosmos.gov.v1beta1.QueryParamsRequest.fromPartial({}));

          case 2:
            response = _context5.sent;
            return _context5.abrupt("return", response);

          case 4:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _query_cosmos_gov_params.apply(this, arguments);
}

function query_cosmos_gov_deposit(_x17, _x18, _x19) {
  return _query_cosmos_gov_deposit.apply(this, arguments);
}

function _query_cosmos_gov_deposit() {
  _query_cosmos_gov_deposit = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(client, proposalId, depositor) {
    var response;
    return _regenerator["default"].wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            _context6.next = 2;
            return client.query.cosmos.gov.v1beta1.deposit(_proto.cosmos.gov.v1beta1.QueryDepositRequest.fromPartial({
              proposalId: proposalId,
              depositor: depositor
            }));

          case 2:
            response = _context6.sent;
            return _context6.abrupt("return", response);

          case 4:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));
  return _query_cosmos_gov_deposit.apply(this, arguments);
}

function query_cosmos_gov_deposits(_x20, _x21, _x22, _x23) {
  return _query_cosmos_gov_deposits.apply(this, arguments);
}

function _query_cosmos_gov_deposits() {
  _query_cosmos_gov_deposits = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(client, proposalId, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee7$(_context7) {
      while (1) {
        switch (_context7.prev = _context7.next) {
          case 0:
            _context7.next = 2;
            return client.query.cosmos.gov.v1beta1.deposits(_proto.cosmos.gov.v1beta1.QueryDepositsRequest.fromPartial({
              proposalId: proposalId,
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context7.sent;
            return _context7.abrupt("return", response);

          case 4:
          case "end":
            return _context7.stop();
        }
      }
    }, _callee7);
  }));
  return _query_cosmos_gov_deposits.apply(this, arguments);
}

function query_cosmos_gov_tally_result(_x24, _x25) {
  return _query_cosmos_gov_tally_result.apply(this, arguments);
} //---------------------------------------------
// COSMOS::GOV MODULE - TRANSACTION HELPERS
// --------------------------------------------
//MsgSubmitProposal


function _query_cosmos_gov_tally_result() {
  _query_cosmos_gov_tally_result = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(client, proposalId) {
    var response;
    return _regenerator["default"].wrap(function _callee8$(_context8) {
      while (1) {
        switch (_context8.prev = _context8.next) {
          case 0:
            _context8.next = 2;
            return client.query.cosmos.gov.v1beta1.tallyResult(_proto.cosmos.gov.v1beta1.QueryTallyResultRequest.fromPartial({
              proposalId: proposalId
            }));

          case 2:
            response = _context8.sent;
            return _context8.abrupt("return", response);

          case 4:
          case "end":
            return _context8.stop();
        }
      }
    }, _callee8);
  }));
  return _query_cosmos_gov_tally_result.apply(this, arguments);
}

function execute_cosmos_gov_submit_proposal_msg(_x26, _x27, _x28) {
  return _execute_cosmos_gov_submit_proposal_msg.apply(this, arguments);
} //MsgVote


function _execute_cosmos_gov_submit_proposal_msg() {
  _execute_cosmos_gov_submit_proposal_msg = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(content, initialDeposit, proposer) {
    var msg;
    return _regenerator["default"].wrap(function _callee9$(_context9) {
      while (1) {
        switch (_context9.prev = _context9.next) {
          case 0:
            msg = _proto.cosmos.gov.v1beta1.MsgSubmitProposal.fromPartial({
              content: content,
              initialDeposit: initialDeposit,
              proposer: proposer
            });
            return _context9.abrupt("return", msg);

          case 2:
          case "end":
            return _context9.stop();
        }
      }
    }, _callee9);
  }));
  return _execute_cosmos_gov_submit_proposal_msg.apply(this, arguments);
}

function execute_cosmos_gov_vote_msg(_x29, _x30, _x31) {
  return _execute_cosmos_gov_vote_msg.apply(this, arguments);
} //MsgVoteWeighted


function _execute_cosmos_gov_vote_msg() {
  _execute_cosmos_gov_vote_msg = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(proposalId, voter, option) {
    var msg;
    return _regenerator["default"].wrap(function _callee10$(_context10) {
      while (1) {
        switch (_context10.prev = _context10.next) {
          case 0:
            msg = _proto.cosmos.gov.v1beta1.MsgVote.fromPartial({
              proposalId: proposalId,
              voter: voter,
              option: option
            });
            return _context10.abrupt("return", msg);

          case 2:
          case "end":
            return _context10.stop();
        }
      }
    }, _callee10);
  }));
  return _execute_cosmos_gov_vote_msg.apply(this, arguments);
}

function execute_cosmos_gov_vote_weighted_msg(_x32, _x33, _x34) {
  return _execute_cosmos_gov_vote_weighted_msg.apply(this, arguments);
} //MsgDeposit


function _execute_cosmos_gov_vote_weighted_msg() {
  _execute_cosmos_gov_vote_weighted_msg = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(proposalId, voter, options) {
    var msg;
    return _regenerator["default"].wrap(function _callee11$(_context11) {
      while (1) {
        switch (_context11.prev = _context11.next) {
          case 0:
            msg = _proto.cosmos.gov.v1beta1.MsgVoteWeighted.fromPartial({
              proposalId: proposalId,
              voter: voter,
              options: options
            });
            return _context11.abrupt("return", msg);

          case 2:
          case "end":
            return _context11.stop();
        }
      }
    }, _callee11);
  }));
  return _execute_cosmos_gov_vote_weighted_msg.apply(this, arguments);
}

function execute_cosmos_gov_deposit_msg(_x35, _x36, _x37) {
  return _execute_cosmos_gov_deposit_msg.apply(this, arguments);
}

function _execute_cosmos_gov_deposit_msg() {
  _execute_cosmos_gov_deposit_msg = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(proposalId, depositor, amount) {
    var msg;
    return _regenerator["default"].wrap(function _callee12$(_context12) {
      while (1) {
        switch (_context12.prev = _context12.next) {
          case 0:
            msg = _proto.cosmos.gov.v1beta1.MsgDeposit.fromPartial({
              proposalId: proposalId,
              depositor: depositor,
              amount: amount
            });
            return _context12.abrupt("return", msg);

          case 2:
          case "end":
            return _context12.stop();
        }
      }
    }, _callee12);
  }));
  return _execute_cosmos_gov_deposit_msg.apply(this, arguments);
}
