// cosmossdkjs/main/utils/cosmos/slashing.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.execute_cosmos_slashing_unjail = execute_cosmos_slashing_unjail;
exports.query_cosmos_slashing_params = query_cosmos_slashing_params;
exports.query_cosmos_slashing_signing_info = query_cosmos_slashing_signing_info;
exports.query_cosmos_slashing_signing_infos = query_cosmos_slashing_signing_infos;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

// Supported query interfaces - types/proto/cosmos/slashing/v1beta1/query.rpc.query.d.ts
// Supported transaction interfaces - types/proto/cosmos/slashing/v1beta1/tx.rpc.msg.d.ts
//---------------------------------------------
// COSMOS::SLASHING MODULE - QUERY HELPERS
// --------------------------------------------
function query_cosmos_slashing_params(_x) {
  return _query_cosmos_slashing_params.apply(this, arguments);
}

function _query_cosmos_slashing_params() {
  _query_cosmos_slashing_params = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client) {
    var response;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return client.query.cosmos.slashing.v1beta1.params(_proto.cosmos.slashing.v1beta1.QueryParamsRequest.fromPartial({}));

          case 2:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _query_cosmos_slashing_params.apply(this, arguments);
}

function query_cosmos_slashing_signing_info(_x2, _x3) {
  return _query_cosmos_slashing_signing_info.apply(this, arguments);
}

function _query_cosmos_slashing_signing_info() {
  _query_cosmos_slashing_signing_info = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client, consAddress) {
    var response;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return client.query.cosmos.slashing.v1beta1.signingInfo(_proto.cosmos.slashing.v1beta1.QuerySigningInfoRequest.fromPartial({
              consAddress: consAddress
            }));

          case 2:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _query_cosmos_slashing_signing_info.apply(this, arguments);
}

function query_cosmos_slashing_signing_infos(_x4, _x5, _x6) {
  return _query_cosmos_slashing_signing_infos.apply(this, arguments);
} //---------------------------------------------
// COSMOS::SLASHING MODULE - TRANSACTION HELPERS
// --------------------------------------------
//MsgUnjail


function _query_cosmos_slashing_signing_infos() {
  _query_cosmos_slashing_signing_infos = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return client.query.cosmos.slashing.v1beta1.signingInfos(_proto.cosmos.slashing.v1beta1.QuerySigningInfosRequest.fromPartial({
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context3.sent;
            return _context3.abrupt("return", response);

          case 4:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _query_cosmos_slashing_signing_infos.apply(this, arguments);
}

function execute_cosmos_slashing_unjail(_x7) {
  return _execute_cosmos_slashing_unjail.apply(this, arguments);
}

function _execute_cosmos_slashing_unjail() {
  _execute_cosmos_slashing_unjail = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(validatorAddr) {
    var msg;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            msg = _proto.cosmos.slashing.v1beta1.MsgUnjail.fromPartial({
              validatorAddr: validatorAddr
            });
            return _context4.abrupt("return", msg);

          case 2:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _execute_cosmos_slashing_unjail.apply(this, arguments);
}
