// cosmossdkjs/main/utils/cosmos/vesting.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.execute_cosmos_vesting_create_vesting_account = execute_cosmos_vesting_create_vesting_account;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

// Supported transaction interfaces - types/proto/cosmos/vesting/v1beta1/tx.rpc.msg.d.ts
//---------------------------------------------
// COSMOS::VESTING MODULE - TRANSACTION HELPERS
// --------------------------------------------
//MsgCreateVestingAccount
function execute_cosmos_vesting_create_vesting_account(_x, _x2, _x3, _x4, _x5) {
  return _execute_cosmos_vesting_create_vesting_account.apply(this, arguments);
}

function _execute_cosmos_vesting_create_vesting_account() {
  _execute_cosmos_vesting_create_vesting_account = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(fromAddress, toAddress, amount, endTime, delayed) {
    var msg;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            msg = _proto.cosmos.vesting.v1beta1.MsgCreateVestingAccount.fromPartial({
              fromAddress: fromAddress,
              toAddress: toAddress,
              amount: amount,
              endTime: endTime,
              delayed: delayed
            });
            return _context.abrupt("return", msg);

          case 2:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _execute_cosmos_vesting_create_vesting_account.apply(this, arguments);
}
