// cosmossdkjs/main/utils/osmosis/incentives.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.execute_osmosis_incentives_add_to_gauge = execute_osmosis_incentives_add_to_gauge;
exports.execute_osmosis_incentives_create_gauge = execute_osmosis_incentives_create_gauge;
exports.query_osmosis_incentives_activeGauges = query_osmosis_incentives_activeGauges;
exports.query_osmosis_incentives_activeGaugesPerDenom = query_osmosis_incentives_activeGaugesPerDenom;
exports.query_osmosis_incentives_gaugeByID = query_osmosis_incentives_gaugeByID;
exports.query_osmosis_incentives_gauges = query_osmosis_incentives_gauges;
exports.query_osmosis_incentives_lockableDurations = query_osmosis_incentives_lockableDurations;
exports.query_osmosis_incentives_moduleDistributedCoins = query_osmosis_incentives_moduleDistributedCoins;
exports.query_osmosis_incentives_moduleToDistributeCoins = query_osmosis_incentives_moduleToDistributeCoins;
exports.query_osmosis_incentives_rewardsEst = query_osmosis_incentives_rewardsEst;
exports.query_osmosis_incentives_upcomingGauges = query_osmosis_incentives_upcomingGauges;
exports.query_osmosis_incentives_upcomingGaugesPerDenom = query_osmosis_incentives_upcomingGaugesPerDenom;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

// Supported query interfaces - types/proto/osmosis/incentives/query.rpc.query.d.ts
// Supported transaction interfaces - types/proto/osmosis/incentives/tx.rpc.msg.d.ts
//---------------------------------------------
// OSMOSIS::INCENTIVES MODULE - QUERY HELPERS
// --------------------------------------------
function query_osmosis_incentives_moduleToDistributeCoins(_x) {
  return _query_osmosis_incentives_moduleToDistributeCoins.apply(this, arguments);
}

function _query_osmosis_incentives_moduleToDistributeCoins() {
  _query_osmosis_incentives_moduleToDistributeCoins = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client) {
    var response;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            console.log(_proto.osmosis.incentives);
            _context.next = 3;
            return client.query.osmosis.incentives.v1.moduleToDistributeCoins(_proto.osmosis.incentives.ModuleToDistributeCoinsRequest.fromPartial({}));

          case 3:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 5:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _query_osmosis_incentives_moduleToDistributeCoins.apply(this, arguments);
}

function query_osmosis_incentives_moduleDistributedCoins(_x2) {
  return _query_osmosis_incentives_moduleDistributedCoins.apply(this, arguments);
}

function _query_osmosis_incentives_moduleDistributedCoins() {
  _query_osmosis_incentives_moduleDistributedCoins = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client) {
    var response;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return client.query.osmosis.incentives.v1.moduleDistributedCoins(_proto.osmosis.incentives.ModuleDistributedCoinsRequest.fromPartial({}));

          case 2:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _query_osmosis_incentives_moduleDistributedCoins.apply(this, arguments);
}

function query_osmosis_incentives_gaugeByID(_x3, _x4) {
  return _query_osmosis_incentives_gaugeByID.apply(this, arguments);
}

function _query_osmosis_incentives_gaugeByID() {
  _query_osmosis_incentives_gaugeByID = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client, id) {
    var response;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return client.query.osmosis.incentives.v1.gaugeByID(_proto.osmosis.incentives.GaugeByIDRequest.fromPartial({
              id: id
            }));

          case 2:
            response = _context3.sent;
            return _context3.abrupt("return", response);

          case 4:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _query_osmosis_incentives_gaugeByID.apply(this, arguments);
}

function query_osmosis_incentives_gauges(_x5, _x6, _x7) {
  return _query_osmosis_incentives_gauges.apply(this, arguments);
}

function _query_osmosis_incentives_gauges() {
  _query_osmosis_incentives_gauges = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(client, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return client.query.osmosis.incentives.v1.gauges(_proto.osmosis.incentives.GaugesRequest.fromPartial({
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context4.sent;
            return _context4.abrupt("return", response);

          case 4:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _query_osmosis_incentives_gauges.apply(this, arguments);
}

function query_osmosis_incentives_activeGauges(_x8, _x9, _x10) {
  return _query_osmosis_incentives_activeGauges.apply(this, arguments);
}

function _query_osmosis_incentives_activeGauges() {
  _query_osmosis_incentives_activeGauges = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(client, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.next = 2;
            return client.query.osmosis.incentives.v1.activeGauges(_proto.osmosis.incentives.ActiveGaugesRequest.fromPartial({
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context5.sent;
            return _context5.abrupt("return", response);

          case 4:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _query_osmosis_incentives_activeGauges.apply(this, arguments);
}

function query_osmosis_incentives_activeGaugesPerDenom(_x11, _x12, _x13, _x14) {
  return _query_osmosis_incentives_activeGaugesPerDenom.apply(this, arguments);
}

function _query_osmosis_incentives_activeGaugesPerDenom() {
  _query_osmosis_incentives_activeGaugesPerDenom = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(client, denom, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            _context6.next = 2;
            return client.query.osmosis.incentives.v1.activeGaugesPerDenom(_proto.osmosis.incentives.ActiveGaugesPerDenomRequest.fromPartial({
              denom: denom,
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context6.sent;
            return _context6.abrupt("return", response);

          case 4:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));
  return _query_osmosis_incentives_activeGaugesPerDenom.apply(this, arguments);
}

function query_osmosis_incentives_upcomingGauges(_x15, _x16, _x17) {
  return _query_osmosis_incentives_upcomingGauges.apply(this, arguments);
}

function _query_osmosis_incentives_upcomingGauges() {
  _query_osmosis_incentives_upcomingGauges = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(client, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee7$(_context7) {
      while (1) {
        switch (_context7.prev = _context7.next) {
          case 0:
            _context7.next = 2;
            return client.query.osmosis.incentives.v1.upcomingGauges(_proto.osmosis.incentives.UpcomingGaugesRequest.fromPartial({
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context7.sent;
            return _context7.abrupt("return", response);

          case 4:
          case "end":
            return _context7.stop();
        }
      }
    }, _callee7);
  }));
  return _query_osmosis_incentives_upcomingGauges.apply(this, arguments);
}

function query_osmosis_incentives_upcomingGaugesPerDenom(_x18, _x19, _x20, _x21) {
  return _query_osmosis_incentives_upcomingGaugesPerDenom.apply(this, arguments);
}

function _query_osmosis_incentives_upcomingGaugesPerDenom() {
  _query_osmosis_incentives_upcomingGaugesPerDenom = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(client, denom, offset, limit) {
    var response;
    return _regenerator["default"].wrap(function _callee8$(_context8) {
      while (1) {
        switch (_context8.prev = _context8.next) {
          case 0:
            _context8.next = 2;
            return client.query.osmosis.incentives.v1.upcomingGaugesPerDenom(_proto.osmosis.incentives.UpcomingGaugesPerDenomRequest.fromPartial({
              denom: denom,
              pagination: _proto.cosmos.base.query.v1beta1.PageRequest.fromPartial({
                offset: offset,
                limit: limit
              })
            }));

          case 2:
            response = _context8.sent;
            return _context8.abrupt("return", response);

          case 4:
          case "end":
            return _context8.stop();
        }
      }
    }, _callee8);
  }));
  return _query_osmosis_incentives_upcomingGaugesPerDenom.apply(this, arguments);
}

function query_osmosis_incentives_rewardsEst(_x22, _x23, _x24, _x25) {
  return _query_osmosis_incentives_rewardsEst.apply(this, arguments);
}

function _query_osmosis_incentives_rewardsEst() {
  _query_osmosis_incentives_rewardsEst = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(client, owner, lockIds, endEpoch) {
    var response;
    return _regenerator["default"].wrap(function _callee9$(_context9) {
      while (1) {
        switch (_context9.prev = _context9.next) {
          case 0:
            _context9.next = 2;
            return client.query.osmosis.incentives.v1.rewardsEst(_proto.osmosis.incentives.RewardsEstRequest.fromPartial({
              owner: owner,
              lockIds: lockIds,
              endEpoch: endEpoch
            }));

          case 2:
            response = _context9.sent;
            return _context9.abrupt("return", response);

          case 4:
          case "end":
            return _context9.stop();
        }
      }
    }, _callee9);
  }));
  return _query_osmosis_incentives_rewardsEst.apply(this, arguments);
}

function query_osmosis_incentives_lockableDurations(_x26) {
  return _query_osmosis_incentives_lockableDurations.apply(this, arguments);
} //---------------------------------------------
// OSMOSIS::INCENTIVES MODULE - TRANSACTION HELPERS
// --------------------------------------------
//MsgCreateGauge


function _query_osmosis_incentives_lockableDurations() {
  _query_osmosis_incentives_lockableDurations = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(client) {
    var response;
    return _regenerator["default"].wrap(function _callee10$(_context10) {
      while (1) {
        switch (_context10.prev = _context10.next) {
          case 0:
            _context10.next = 2;
            return client.query.osmosis.incentives.v1.lockableDurations(_proto.osmosis.incentives.QueryLockableDurationsRequest.fromPartial({}));

          case 2:
            response = _context10.sent;
            return _context10.abrupt("return", response);

          case 4:
          case "end":
            return _context10.stop();
        }
      }
    }, _callee10);
  }));
  return _query_osmosis_incentives_lockableDurations.apply(this, arguments);
}

function execute_osmosis_incentives_create_gauge(_x27, _x28, _x29, _x30, _x31, _x32) {
  return _execute_osmosis_incentives_create_gauge.apply(this, arguments);
} //MsgAddToGauge


function _execute_osmosis_incentives_create_gauge() {
  _execute_osmosis_incentives_create_gauge = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(isPerpetual, owner, distributeTo, coins, startTime, numEpochsPaidOver) {
    var msg;
    return _regenerator["default"].wrap(function _callee11$(_context11) {
      while (1) {
        switch (_context11.prev = _context11.next) {
          case 0:
            msg = _proto.osmosis.incentives.MsgCreateGauge.fromPartial({
              isPerpetual: isPerpetual,
              owner: owner,
              distributeTo: distributeTo,
              coins: coins,
              startTime: startTime,
              numEpochsPaidOver: numEpochsPaidOver
            });
            return _context11.abrupt("return", msg);

          case 2:
          case "end":
            return _context11.stop();
        }
      }
    }, _callee11);
  }));
  return _execute_osmosis_incentives_create_gauge.apply(this, arguments);
}

function execute_osmosis_incentives_add_to_gauge(_x33, _x34, _x35) {
  return _execute_osmosis_incentives_add_to_gauge.apply(this, arguments);
}

function _execute_osmosis_incentives_add_to_gauge() {
  _execute_osmosis_incentives_add_to_gauge = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(owner, gaugeId, rewards) {
    var msg;
    return _regenerator["default"].wrap(function _callee12$(_context12) {
      while (1) {
        switch (_context12.prev = _context12.next) {
          case 0:
            msg = _proto.osmosis.incentives.MsgAddToGauge.fromPartial({
              owner: owner,
              gaugeId: gaugeId,
              rewards: rewards
            });
            return _context12.abrupt("return", msg);

          case 2:
          case "end":
            return _context12.stop();
        }
      }
    }, _callee12);
  }));
  return _execute_osmosis_incentives_add_to_gauge.apply(this, arguments);
}
