// cosmossdkjs/main/utils/osmosis/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "execute_osmosis_gamm_ExitPool", {
  enumerable: true,
  get: function get() {
    return _gamm.execute_osmosis_gamm_ExitPool;
  }
});
Object.defineProperty(exports, "execute_osmosis_gamm_JoinPool", {
  enumerable: true,
  get: function get() {
    return _gamm.execute_osmosis_gamm_JoinPool;
  }
});
Object.defineProperty(exports, "execute_osmosis_gamm_SwapExactAmountIn", {
  enumerable: true,
  get: function get() {
    return _gamm.execute_osmosis_gamm_SwapExactAmountIn;
  }
});
Object.defineProperty(exports, "execute_osmosis_gamm_SwapExactAmountOut", {
  enumerable: true,
  get: function get() {
    return _gamm.execute_osmosis_gamm_SwapExactAmountOut;
  }
});
Object.defineProperty(exports, "execute_osmosis_gamm_exitSwapExternAmountOut", {
  enumerable: true,
  get: function get() {
    return _gamm.execute_osmosis_gamm_exitSwapExternAmountOut;
  }
});
Object.defineProperty(exports, "execute_osmosis_gamm_exitSwapShareAmountIn", {
  enumerable: true,
  get: function get() {
    return _gamm.execute_osmosis_gamm_exitSwapShareAmountIn;
  }
});
Object.defineProperty(exports, "execute_osmosis_gamm_joinSwapExternAmountIn", {
  enumerable: true,
  get: function get() {
    return _gamm.execute_osmosis_gamm_joinSwapExternAmountIn;
  }
});
Object.defineProperty(exports, "execute_osmosis_gamm_joinSwapShareAmountOut", {
  enumerable: true,
  get: function get() {
    return _gamm.execute_osmosis_gamm_joinSwapShareAmountOut;
  }
});
Object.defineProperty(exports, "query_osmosis_epoch_currentEpoch", {
  enumerable: true,
  get: function get() {
    return _epochs.query_osmosis_epoch_currentEpoch;
  }
});
Object.defineProperty(exports, "query_osmosis_epoch_epochInfos", {
  enumerable: true,
  get: function get() {
    return _epochs.query_osmosis_epoch_epochInfos;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_estimateSwapExactAmountIn", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_estimateSwapExactAmountIn;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_estimateSwapExactAmountOut", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_estimateSwapExactAmountOut;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_numpools", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_numpools;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_pool", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_pool;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_poolParams", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_poolParams;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_pools", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_pools;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_spotPrice", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_spotPrice;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_totalLiquidity", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_totalLiquidity;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_totalPoolLiquidity", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_totalPoolLiquidity;
  }
});
Object.defineProperty(exports, "query_osmosis_gamm_totalShares", {
  enumerable: true,
  get: function get() {
    return _gamm.query_osmosis_gamm_totalShares;
  }
});

var _epochs = require("./epochs");

var _gamm = require("./gamm");
