// cosmossdkjs/main/utils/osmosis/pool_incentives.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.query_osmosis_pool_incentives_distrInfo = query_osmosis_pool_incentives_distrInfo;
exports.query_osmosis_pool_incentives_externalIncentiveGauges = query_osmosis_pool_incentives_externalIncentiveGauges;
exports.query_osmosis_pool_incentives_gaugeIds = query_osmosis_pool_incentives_gaugeIds;
exports.query_osmosis_pool_incentives_incentivizedPools = query_osmosis_pool_incentives_incentivizedPools;
exports.query_osmosis_pool_incentives_lockableDurations = query_osmosis_pool_incentives_lockableDurations;
exports.query_osmosis_pool_incentives_params = query_osmosis_pool_incentives_params;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _proto = require("../../proto");

// Supported query interfaces - types/proto/osmosis/pool-incentives/v1beta1/query.rpc.query.d.ts
// Supported transaction interfaces - types/proto/osmosis/pool-incentives/v1beta1/superfactory/tx.rpc.msg.d.ts
//---------------------------------------------
// OSMOSIS::POOL-INCENTIVES MODULE - QUERY HELPERS
// --------------------------------------------
function query_osmosis_pool_incentives_gaugeIds(_x, _x2) {
  return _query_osmosis_pool_incentives_gaugeIds.apply(this, arguments);
}

function _query_osmosis_pool_incentives_gaugeIds() {
  _query_osmosis_pool_incentives_gaugeIds = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(client, poolId) {
    var response;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return client.query.osmosis.pool_incentives.v1beta1.gaugeIds(_proto.osmosis.poolincentives.v1beta1.QueryGaugeIdsRequest.fromPartial({
              poolId: poolId
            }));

          case 2:
            response = _context.sent;
            return _context.abrupt("return", response);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _query_osmosis_pool_incentives_gaugeIds.apply(this, arguments);
}

function query_osmosis_pool_incentives_distrInfo(_x3) {
  return _query_osmosis_pool_incentives_distrInfo.apply(this, arguments);
}

function _query_osmosis_pool_incentives_distrInfo() {
  _query_osmosis_pool_incentives_distrInfo = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(client) {
    var response;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return client.query.osmosis.pool_incentives.v1beta1.distrInfo(_proto.osmosis.poolincentives.v1beta1.QueryGaugeIdsRequest.fromPartial({}));

          case 2:
            response = _context2.sent;
            return _context2.abrupt("return", response);

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _query_osmosis_pool_incentives_distrInfo.apply(this, arguments);
}

function query_osmosis_pool_incentives_params(_x4) {
  return _query_osmosis_pool_incentives_params.apply(this, arguments);
}

function _query_osmosis_pool_incentives_params() {
  _query_osmosis_pool_incentives_params = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(client) {
    var response;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return client.query.osmosis.pool_incentives.v1beta1.params(_proto.osmosis.poolincentives.v1beta1.QueryGaugeIdsRequest.fromPartial({}));

          case 2:
            response = _context3.sent;
            return _context3.abrupt("return", response);

          case 4:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _query_osmosis_pool_incentives_params.apply(this, arguments);
}

function query_osmosis_pool_incentives_lockableDurations(_x5) {
  return _query_osmosis_pool_incentives_lockableDurations.apply(this, arguments);
}

function _query_osmosis_pool_incentives_lockableDurations() {
  _query_osmosis_pool_incentives_lockableDurations = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(client) {
    var response;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return client.query.osmosis.pool_incentives.v1beta1.lockableDurations(_proto.osmosis.poolincentives.v1beta1.QueryGaugeIdsRequest.fromPartial({}));

          case 2:
            response = _context4.sent;
            return _context4.abrupt("return", response);

          case 4:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _query_osmosis_pool_incentives_lockableDurations.apply(this, arguments);
}

function query_osmosis_pool_incentives_incentivizedPools(_x6) {
  return _query_osmosis_pool_incentives_incentivizedPools.apply(this, arguments);
}

function _query_osmosis_pool_incentives_incentivizedPools() {
  _query_osmosis_pool_incentives_incentivizedPools = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(client) {
    var response;
    return _regenerator["default"].wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.next = 2;
            return client.query.osmosis.pool_incentives.v1beta1.incentivizedPools(_proto.osmosis.poolincentives.v1beta1.QueryGaugeIdsRequest.fromPartial({}));

          case 2:
            response = _context5.sent;
            return _context5.abrupt("return", response);

          case 4:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _query_osmosis_pool_incentives_incentivizedPools.apply(this, arguments);
}

function query_osmosis_pool_incentives_externalIncentiveGauges(_x7) {
  return _query_osmosis_pool_incentives_externalIncentiveGauges.apply(this, arguments);
}

function _query_osmosis_pool_incentives_externalIncentiveGauges() {
  _query_osmosis_pool_incentives_externalIncentiveGauges = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(client) {
    var response;
    return _regenerator["default"].wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            _context6.next = 2;
            return client.query.osmosis.pool_incentives.v1beta1.externalIncentiveGauges(_proto.osmosis.poolincentives.v1beta1.QueryGaugeIdsRequest.fromPartial({}));

          case 2:
            response = _context6.sent;
            return _context6.abrupt("return", response);

          case 4:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));
  return _query_osmosis_pool_incentives_externalIncentiveGauges.apply(this, arguments);
}
