// @cosmos-client/core/esm/rest/gov/module.js
import { QueryApi } from '../../openapi/api';
// deprecated gov/v1beta1
export function params(sdk, paramsType) {
    return new QueryApi(undefined, sdk.url).govParams(paramsType);
}
export function govV1Params(sdk, paramsType) {
    return new QueryApi(undefined, sdk.url).govV1Params(paramsType);
}
// deprecated gov/v1beta1
export function proposals(sdk, proposalStatus, voter, depositor, paginationKey, paginationOffset, paginationLimit, paginationCountTotal) {
    return new QueryApi(undefined, sdk.url).proposals(proposalStatus, voter, depositor, paginationKey, paginationOffset?.toString(), paginationLimit?.toString(), paginationCountTotal);
}
export function govV1Proposals(sdk, proposalStatus, voter, depositor, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse) {
    return new QueryApi(undefined, sdk.url).govV1Proposal(proposalStatus, voter, depositor, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse);
}
// deprecated gov/v1beta1
export function proposal(sdk, proposalID) {
    return new QueryApi(undefined, sdk.url).proposal(proposalID);
}
export function govV1Proposal(sdk, proposalID) {
    return new QueryApi(undefined, sdk.url).govV1Proposal_2(proposalID);
}
// deprecated gov/v1beta1
export function deposits(sdk, proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal) {
    return new QueryApi(undefined, sdk.url).deposits(proposalID, paginationKey, paginationOffset?.toString(), paginationLimit?.toString(), paginationCountTotal);
}
export function govV1Deposits(sdk, proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse) {
    return new QueryApi(undefined, sdk.url).govV1Deposit(proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse);
}
// deprecated gov/v1beta1
export function deposit(sdk, proposalID, depositor) {
    return new QueryApi(undefined, sdk.url).deposit(proposalID, depositor);
}
export function govV1Deposit(sdk, proposalID, depositor) {
    return new QueryApi(undefined, sdk.url).govV1Deposit_1(proposalID, depositor);
}
// deprecated gov/v1beta1
export function tallyresult(sdk, proposalID) {
    return new QueryApi(undefined, sdk.url).tallyResult(proposalID);
}
export function govV1TallyResult(sdk, proposalID) {
    return new QueryApi(undefined, sdk.url).govV1TallyResult(proposalID);
}
// deprecated gov/v1beta1
export function votes(sdk, proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal) {
    return new QueryApi(undefined, sdk.url).votes(proposalID, paginationKey, paginationOffset?.toString(), paginationLimit?.toString(), paginationCountTotal);
}
export function govV1Votes(sdk, proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse) {
    return new QueryApi(undefined, sdk.url).govV1Votes(proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse);
}
// deprecated gov/v1beta1
export function vote(sdk, proposalID, voter) {
    return new QueryApi(undefined, sdk.url).vote(proposalID, voter);
}
export function govV1Vote(sdk, proposalID, voter) {
    return new QueryApi(undefined, sdk.url).govV1Vote(proposalID, voter);
}
