// chain-registry/main/mainnet/gitopia/assets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var info = {
  $schema: '../assetlist.schema.json',
  chain_name: 'gitopia',
  assets: [{
    description: 'The native token of Gitopia',
    denom_units: [{
      denom: 'ulore',
      exponent: 0
    }, {
      denom: 'LORE',
      exponent: 6
    }],
    base: 'ulore',
    name: 'LORE',
    display: 'LORE',
    symbol: 'LORE',
    logo_URIs: {
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/gitopia/images/lore.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/gitopia/images/lore.svg'
    },
    images: [{
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/gitopia/images/lore.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/gitopia/images/lore.svg'
    }, {
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/gitopia/images/gitopia.png'
    }],
    coingecko_id: 'gitopia'
  }]
};
var _default = exports["default"] = info;
