// chain-registry/main/testnet/rsprovidertestnet/assets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var info = {
  $schema: '../../assetlist.schema.json',
  chain_name: 'rsprovidertestnet',
  assets: [{
    description: 'The native staking and governance token of the Replicated Security Testnet.',
    denom_units: [{
      denom: 'uatom',
      exponent: 0
    }, {
      denom: 'atom',
      exponent: 6
    }],
    base: 'uatom',
    name: 'Cosmos',
    display: 'atom',
    symbol: 'ATOM',
    logo_URIs: {
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/cosmoshub/images/atom.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/cosmoshub/images/atom.svg'
    },
    images: [{
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/cosmoshub/images/atom.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/cosmoshub/images/atom.svg'
    }]
  }]
};
var _default = exports["default"] = info;
