// cosmossdkjs/main/utils/cosmwasm/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "find_code_id_from_contract_hash", {
  enumerable: true,
  get: function get() {
    return _wasm.find_code_id_from_contract_hash;
  }
});
Object.defineProperty(exports, "make_clearAdmin_msg", {
  enumerable: true,
  get: function get() {
    return _wasm.make_clearAdmin_msg;
  }
});
Object.defineProperty(exports, "make_executeContract_msg", {
  enumerable: true,
  get: function get() {
    return _wasm.make_executeContract_msg;
  }
});
Object.defineProperty(exports, "make_instantiateContract_msg", {
  enumerable: true,
  get: function get() {
    return _wasm.make_instantiateContract_msg;
  }
});
Object.defineProperty(exports, "make_migrateContract_msg", {
  enumerable: true,
  get: function get() {
    return _wasm.make_migrateContract_msg;
  }
});
Object.defineProperty(exports, "make_storeCode_msg", {
  enumerable: true,
  get: function get() {
    return _wasm.make_storeCode_msg;
  }
});
Object.defineProperty(exports, "make_updateAdmin_msg", {
  enumerable: true,
  get: function get() {
    return _wasm.make_updateAdmin_msg;
  }
});
Object.defineProperty(exports, "query_wasm_allContractState", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_allContractState;
  }
});
Object.defineProperty(exports, "query_wasm_code", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_code;
  }
});
Object.defineProperty(exports, "query_wasm_codes", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_codes;
  }
});
Object.defineProperty(exports, "query_wasm_contractHistory", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_contractHistory;
  }
});
Object.defineProperty(exports, "query_wasm_contractInfo", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_contractInfo;
  }
});
Object.defineProperty(exports, "query_wasm_contractsByCode", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_contractsByCode;
  }
});
Object.defineProperty(exports, "query_wasm_pinnedCodes", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_pinnedCodes;
  }
});
Object.defineProperty(exports, "query_wasm_rawContractState", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_rawContractState;
  }
});
Object.defineProperty(exports, "query_wasm_smartContractState", {
  enumerable: true,
  get: function get() {
    return _wasm.query_wasm_smartContractState;
  }
});

var _wasm = require("./wasm");
