// cosmossdkjs/module/proto/osmosis/tokenfactory/v1beta1/authorityMetadata.js
import * as _m0 from "protobufjs/minimal";
import { isSet } from "@osmonauts/helpers";
/**
 * DenomAuthorityMetadata specifies metadata for addresses that have specific
 * capabilities over a token factory denom. Right now there is only one Admin
 * permission, but is planned to be extended to the future.
 */

function createBaseDenomAuthorityMetadata() {
  return {
    Admin: ""
  };
}

export const DenomAuthorityMetadata = {
  encode(message, writer = _m0.Writer.create()) {
    if (message.Admin !== "") {
      writer.uint32(10).string(message.Admin);
    }

    return writer;
  },

  decode(input, length) {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseDenomAuthorityMetadata();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.Admin = reader.string();
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object) {
    return {
      Admin: isSet(object.Admin) ? String(object.Admin) : ""
    };
  },

  toJSON(message) {
    const obj = {};
    message.Admin !== undefined && (obj.Admin = message.Admin);
    return obj;
  },

  fromPartial(object) {
    const message = createBaseDenomAuthorityMetadata();
    message.Admin = object.Admin ?? "";
    return message;
  }

};
