// chain-registry/main/mainnet/aioz/assets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var info = {
  $schema: '../assetlist.schema.json',
  chain_name: 'aioz',
  assets: [{
    description: 'The native staking and governance token of the AIOZ Network.',
    denom_units: [{
      denom: 'attoaioz',
      exponent: 0
    }, {
      denom: 'nanoaioz',
      exponent: 9
    }, {
      denom: 'aioz',
      exponent: 18
    }],
    base: 'attoaioz',
    name: 'AIOZ',
    display: 'aioz',
    symbol: 'AIOZ',
    logo_URIs: {
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/aioz/images/aioz.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/aioz/images/aioz.svg'
    },
    coingecko_id: 'aioz-network',
    images: [{
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/aioz/images/aioz.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/aioz/images/aioz.svg'
    }]
  }]
};
var _default = exports["default"] = info;
