// @cosmos-client/core/cjs/rest/gov/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.govV1Vote = exports.vote = exports.govV1Votes = exports.votes = exports.govV1TallyResult = exports.tallyresult = exports.govV1Deposit = exports.deposit = exports.govV1Deposits = exports.deposits = exports.govV1Proposal = exports.proposal = exports.govV1Proposals = exports.proposals = exports.govV1Params = exports.params = void 0;
var api_1 = require("../../openapi/api");
// deprecated gov/v1beta1
function params(sdk, paramsType) {
    return new api_1.QueryApi(undefined, sdk.url).govParams(paramsType);
}
exports.params = params;
function govV1Params(sdk, paramsType) {
    return new api_1.QueryApi(undefined, sdk.url).govV1Params(paramsType);
}
exports.govV1Params = govV1Params;
// deprecated gov/v1beta1
function proposals(sdk, proposalStatus, voter, depositor, paginationKey, paginationOffset, paginationLimit, paginationCountTotal) {
    return new api_1.QueryApi(undefined, sdk.url).proposals(proposalStatus, voter, depositor, paginationKey, paginationOffset === null || paginationOffset === void 0 ? void 0 : paginationOffset.toString(), paginationLimit === null || paginationLimit === void 0 ? void 0 : paginationLimit.toString(), paginationCountTotal);
}
exports.proposals = proposals;
function govV1Proposals(sdk, proposalStatus, voter, depositor, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse) {
    return new api_1.QueryApi(undefined, sdk.url).govV1Proposal(proposalStatus, voter, depositor, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse);
}
exports.govV1Proposals = govV1Proposals;
// deprecated gov/v1beta1
function proposal(sdk, proposalID) {
    return new api_1.QueryApi(undefined, sdk.url).proposal(proposalID);
}
exports.proposal = proposal;
function govV1Proposal(sdk, proposalID) {
    return new api_1.QueryApi(undefined, sdk.url).govV1Proposal_2(proposalID);
}
exports.govV1Proposal = govV1Proposal;
// deprecated gov/v1beta1
function deposits(sdk, proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal) {
    return new api_1.QueryApi(undefined, sdk.url).deposits(proposalID, paginationKey, paginationOffset === null || paginationOffset === void 0 ? void 0 : paginationOffset.toString(), paginationLimit === null || paginationLimit === void 0 ? void 0 : paginationLimit.toString(), paginationCountTotal);
}
exports.deposits = deposits;
function govV1Deposits(sdk, proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse) {
    return new api_1.QueryApi(undefined, sdk.url).govV1Deposit(proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse);
}
exports.govV1Deposits = govV1Deposits;
// deprecated gov/v1beta1
function deposit(sdk, proposalID, depositor) {
    return new api_1.QueryApi(undefined, sdk.url).deposit(proposalID, depositor);
}
exports.deposit = deposit;
function govV1Deposit(sdk, proposalID, depositor) {
    return new api_1.QueryApi(undefined, sdk.url).govV1Deposit_1(proposalID, depositor);
}
exports.govV1Deposit = govV1Deposit;
// deprecated gov/v1beta1
function tallyresult(sdk, proposalID) {
    return new api_1.QueryApi(undefined, sdk.url).tallyResult(proposalID);
}
exports.tallyresult = tallyresult;
function govV1TallyResult(sdk, proposalID) {
    return new api_1.QueryApi(undefined, sdk.url).govV1TallyResult(proposalID);
}
exports.govV1TallyResult = govV1TallyResult;
// deprecated gov/v1beta1
function votes(sdk, proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal) {
    return new api_1.QueryApi(undefined, sdk.url).votes(proposalID, paginationKey, paginationOffset === null || paginationOffset === void 0 ? void 0 : paginationOffset.toString(), paginationLimit === null || paginationLimit === void 0 ? void 0 : paginationLimit.toString(), paginationCountTotal);
}
exports.votes = votes;
function govV1Votes(sdk, proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse) {
    return new api_1.QueryApi(undefined, sdk.url).govV1Votes(proposalID, paginationKey, paginationOffset, paginationLimit, paginationCountTotal, paginationReverse);
}
exports.govV1Votes = govV1Votes;
// deprecated gov/v1beta1
function vote(sdk, proposalID, voter) {
    return new api_1.QueryApi(undefined, sdk.url).vote(proposalID, voter);
}
exports.vote = vote;
function govV1Vote(sdk, proposalID, voter) {
    return new api_1.QueryApi(undefined, sdk.url).govV1Vote(proposalID, voter);
}
exports.govV1Vote = govV1Vote;
