// chain-registry/main/mainnet/planq/assets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var info = {
  $schema: '../assetlist.schema.json',
  chain_name: 'planq',
  assets: [{
    description: 'The native EVM, governance and staking token of the Planq Network',
    denom_units: [{
      denom: 'aplanq',
      exponent: 0
    }, {
      denom: 'planq',
      exponent: 18
    }],
    base: 'aplanq',
    name: 'Planq',
    display: 'planq',
    symbol: 'PLQ',
    logo_URIs: {
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/planq/images/planq.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/planq/images/planq.svg'
    },
    coingecko_id: 'planq',
    images: [{
      png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/planq/images/planq.png',
      svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/planq/images/planq.svg'
    }]
  }]
};
var _default = exports["default"] = info;
